python/arcticpy/read_noise.py

```diff
@@ -232,17 +239,154 @@ def estimate_residual_covariance(
+        #add read noise
+        noiseFrame = np.random.normal(0, self.sigmaRN, (chip_size,chip_size))
+        readNoiseAddedFrame = ctiTrailedFrame + noiseFrame
+        #do S+R routine
+        sFrame, rFrame = self.estimate_read_noise_model_from_image(readNoiseAddedFrame)
+        #clean CTI from S frame
+        ctiCorrectedFrame = ac.remove_cti(sFrame,1,parallel_roe=parallel_roe,parallel_ccd=parallel_ccd,parallel_traps=parallel_traps,parallel_express=parallel_express,serial_roe=serial_roe,serial_ccd=serial_ccd,serial_traps=serial_traps,serial_express=serial_express)
+        #re-add R frame to correction
+        outputFrame = ctiCorrectedFrame + rFrame
+
+        #determine covariance FOMs
+        fomSR = self.covariance_matrix_from_image(outputFrame,matrix_size=matrix_size)
+        fomBenchmark = self.covariance_matrix_from_image(skyFrame+noiseFrame,matrix_size=matrix_size)
+        fomDiff = fomSR - fomBenchmark
+        
+        return fomDiff
+        
+        #raise NotImplementedError
+        #return np.array([[0,0,0,0,0],[0,0,0,0,0],[0,0,1,0,0],[0,0,0,0,0],[0,0,0,0,0]])
+
+    ###############
+    ###############
+    def _estimate_residual_covariance_for_opt(
+            self, 
+            background_level,
+            sr_fraction,
+            background_sigma=None,
+            chip_size=500,  #size, in pixels, of CCD (assumes square)
+            matrix_size=5, #size of covariance grid, in pixels (also assumes square)
+            **kwargs
+            ### Parallel
+            ##parallel_ccd=None, #can we do this with **kwargs ? I'm not sure in python
+            ##parallel_roe=None,
+            ##parallel_traps=None,
+            ##parallel_express=0,
+            ##parallel_window_offset=0,
+            ##parallel_window_start=0,
+            ##parallel_window_stop=-1,
+            ##parallel_time_start=0,
+            ##parallel_time_stop=-1,
+            ##parallel_prune_n_electrons=1e-10, 
+            ##parallel_prune_frequency=20,
+            ### Serial  
+            ##serial_ccd=None,
+            ##serial_roe=None,
+            ##serial_traps=None,
+            ##serial_express=0,
+            ##serial_window_offset=0,
+            ##serial_window_start=0,
+            ##serial_window_stop=-1,
+            ##serial_time_start=0,
+            ##serial_time_stop=-1,
+            ##serial_prune_n_electrons=1e-10, 
+            ##serial_prune_frequency=20,
+            ### Pixel bounce
+            ##pixel_bounce=None,
+            ### Output
+            ##verbosity=1,
+    ):
+        if background_sigma==None:
+            background_sigma = np.sqrt(background_level) #if sky sigma is not set, assume sqrt(N)
+
+        if 'parallel_roe' in kwargs:
+            parallel_roe = kwargs['parallel_roe']
```

Comment:
You can replace all of these if statements with:

`parallel_roe = kwargs.get('parallel_roe')`

This will return None if the key is not in the dictionary.